File: server/analyticsManager.js
// analyticsManager.js

const fs = require("fs");
const path = require("path");
const fetch = require("node-fetch");
const { Wallet, SecretNetworkClient } = require("secretjs");

// Use your token details from the front-end tokens file
const tokens = {
  ERTH: {
    contract: "secret16snu3lt8k9u0xr54j2hqyhvwnx9my7kq7ay8lp",
    hash: "638a3e1d50175fbcb8373cf801565283e3eb23d88a9b7b7f99fcc5eb1e6b561e",
    decimals: 6,
    logo: "/images/logo.png",
  },
  ANML: {
    contract: "secret14p6dhjznntlzw0yysl7p6z069nk0skv5e9qjut",
    hash: "638a3e1d50175fbcb8373cf801565283e3eb23d88a9b7b7f99fcc5eb1e6b561e",
    decimals: 6,
    logo: "/images/anml.png",
    lp: {
      contract: "secret1ztcedff57xqmt4lwnpdrz865rtrst037n6d8tq",
      hash: "638a3e1d50175fbcb8373cf801565283e3eb23d88a9b7b7f99fcc5eb1e6b561e",
      decimals: 6,
    },
  },
  FINA: {
    contract: "secret1s3z9xkpdsrhk86300tqnv6u466jmdmlegew2ve",
    hash: "cfecd51a022c520c55429d974363fd7f065d20474af6a362da8737f73b7d9e80",
    decimals: 6,
    logo: "/images/coin/FINA.webp",
    coingeckoId: "fina",
    lp: {
      contract: "secret14ttyephg7pvzzt0n53y888rvmeec5hl0pnf5tw",
      hash: "638a3e1d50175fbcb8373cf801565283e3eb23d88a9b7b7f99fcc5eb1e6b561e",
      decimals: 6,
    },
  },
  sSCRT: {
    contract: "secret1k0jntykt7e4g3y88ltc60czgjuqdy4c9e8fzek",
    hash: "af74387e276be8874f07bec3a87023ee49b0e7ebe08178c49d0a49c3c98ed60e",
    decimals: 6,
    logo: "/images/coin/SSCRT.png",
    coingeckoId: "secret",
    lp: {
      contract: "secret1hk5lj62949s88q7kl892rkxyhxcxdpznzlw924",
      hash: "638a3e1d50175fbcb8373cf801565283e3eb23d88a9b7b7f99fcc5eb1e6b561e",
      decimals: 6,
    },
  },
};

// Where to store persistent data
const ANALYTICS_FILE = path.join(__dirname, "analyticsData.json");

// Utility function to read file contents
function get_value(file) {
  const filePath = path.join(__dirname, file);
  try {
    const data = fs.readFileSync(filePath, "utf8");
    return data.trim(); // Trim to remove any extra whitespace or newlines
  } catch (err) {
    console.error(err);
    return null; // Handle the error as needed
  }
}

// In-memory history
let analyticsHistory = [];
const WALLET_KEY = get_value("WALLET_KEY.txt");
// Initialize wallet and Secret Network client
const wallet = new Wallet(WALLET_KEY);
const secretjs = new SecretNetworkClient({
  url: "https://lcd.erth.network",
  chainId: "secret-4",
  wallet: wallet,
  walletAddress: wallet.address,
});

// Load analytics data from file
function loadAnalyticsData() {
  try {
    if (fs.existsSync(ANALYTICS_FILE)) {
      const raw = fs.readFileSync(ANALYTICS_FILE, "utf8");
      analyticsHistory = JSON.parse(raw);
      console.log(`[analyticsManager] Loaded ${analyticsHistory.length} historical data points`);
    }
  } catch (err) {
    console.error("Error loading analytics data:", err);
  }
}

// Save analytics data to file
function saveAnalyticsData() {
  fs.writeFileSync(ANALYTICS_FILE, JSON.stringify(analyticsHistory, null, 2), "utf8");
}

// Reset analytics data
function resetAnalyticsData() {
  console.log("[analyticsManager] Resetting analytics data");
  analyticsHistory = [];
  saveAnalyticsData();
  console.log("[analyticsManager] Analytics data has been reset");
}

// Add unified pool query constants
const UNIFIED_POOL_CONTRACT = "secret1rj2phrf6x3v7526jrz60m2dcq58slyq2269kra"; // Exchange contract
const UNIFIED_POOL_HASH = "1c2220105c2a33edf4bbafacecb6cbdf317dac26289ada1df0cec1abc73895bd"; // Correct exchange contract hash

async function updateErthValues() {
  try {
    console.log("[analyticsManager] Updating ERTH analytics...");

    // 1) Fetch token prices from Coingecko (for tokens with coingeckoId)
    const tokenIds = Object.values(tokens)
      .filter((t) => t.coingeckoId)
      .map((t) => t.coingeckoId)
      .join(",");
    const priceRes = await fetch(`https://api.coingecko.com/api/v3/simple/price?ids=${tokenIds}&vs_currencies=usd`);
    const priceData = await priceRes.json();
    console.log("[DEBUG] Coingecko priceData:", priceData);
    const prices = {};
    for (const k in tokens) {
      const t = tokens[k];
      if (t.coingeckoId && priceData[t.coingeckoId]) {
        prices[k] = priceData[t.coingeckoId].usd;
      }
    }
    console.log("[DEBUG] Computed prices:", prices);

    // 2) Query ERTH total supply
    const erthInfo = await secretjs.query.compute.queryContract({
      contract_address: tokens.ERTH.contract,
      code_hash: tokens.ERTH.hash,
      query: { token_info: {} },
    });
    const erthTotalSupply = parseInt(erthInfo.token_info.total_supply) / Math.pow(10, tokens.ERTH.decimals);
    console.log("[DEBUG] ERTH total supply:", erthTotalSupply);

    // 3) Query ANML total supply
    const anmlInfo = await secretjs.query.compute.queryContract({
      contract_address: tokens.ANML.contract,
      code_hash: tokens.ANML.hash,
      query: { token_info: {} },
    });
    const anmlTotalSupply = parseInt(anmlInfo.token_info.total_supply) / Math.pow(10, tokens.ANML.decimals);
    console.log("[DEBUG] ANML total supply:", anmlTotalSupply);

    // 4) Build pool addresses list using each token's contract (for pools)
    const poolQueryTokens = [];
    const poolAddresses = [];
    for (const key in tokens) {
      const tk = tokens[key];
      // For pools we require a price (even if ANML doesn't have one, we check for its existence separately)
      if (key !== "ERTH" && tk.contract && (prices[key] !== undefined || key === "ANML")) {
        poolQueryTokens.push({ tokenKey: key, decimals: tk.decimals });
        poolAddresses.push(tk.contract);
      }
    }
    console.log("[DEBUG] Pool addresses:", poolAddresses);

    // 5) Unified query to fetch all pool info
    const unifiedPoolRes = await secretjs.query.compute.queryContract({
      contract_address: UNIFIED_POOL_CONTRACT,
      code_hash: UNIFIED_POOL_HASH,
      query: { query_pool_info: { pools: poolAddresses } },
    });
    console.log("[DEBUG] Unified pool response:", unifiedPoolRes);

    let totalWeightedPrice = 0;
    let totalLiquidity = 0;
    let poolData = [];
    let anmlPriceFinal = null;
    let anmlTVL = 0;
    let anmlData = null;

    // Process each pool
    for (let i = 0; i < unifiedPoolRes.length; i++) {
      const st = unifiedPoolRes[i];
      const tokenKey = poolQueryTokens[i].tokenKey;
      const tk = tokens[tokenKey];
      const erthReserve = parseInt(st.state.erth_reserve) / Math.pow(10, tokens.ERTH.decimals);
      const tokenReserve = parseInt(st.state.token_b_reserve) / Math.pow(10, tk.decimals);
      console.log(`[DEBUG] Pool ${tokenKey}: erthReserve=${erthReserve}, tokenReserve=${tokenReserve}`);
      if (erthReserve === 0) {
        console.error(`[ERROR] Zero ERTH reserve for pool ${tokenKey}`);
        continue;
      }
      if (tokenKey === "ANML") {
        // Store ANML pool data for later processing (ANML doesn't have a Coingecko price)
        anmlData = { tokenKey, tokenReserve, erthReserve };
      } else {
        // For non‑ANML pools, compute pool price from external token price
        const poolPrice = (tokenReserve / erthReserve) * prices[tokenKey];
        const poolTVL = tokenReserve * prices[tokenKey] + erthReserve * poolPrice;
        totalWeightedPrice += poolPrice * poolTVL;
        totalLiquidity += poolTVL;
        poolData.push({ token: tokenKey, erthPrice: poolPrice, tvl: poolTVL });
      }
    }

    // 6) Compute global ERTH price from non‑ANML pools
    const globalErthPrice = totalLiquidity ? totalWeightedPrice / totalLiquidity : 0;
    console.log("[DEBUG] Global ERTH price:", globalErthPrice);

    // 7) Process ANML pool using its reserves and the global ERTH price
    let anmlMarketCap = 0;
    if (anmlData) {
      // Derive ANML price from pool reserves: (ERTH reserve / ANML reserve) * global ERTH price
      anmlPriceFinal = (anmlData.erthReserve / anmlData.tokenReserve) * globalErthPrice;
      anmlTVL = anmlData.tokenReserve * anmlPriceFinal + anmlData.erthReserve * globalErthPrice;
      anmlMarketCap = anmlPriceFinal * anmlTotalSupply;
      totalLiquidity += anmlTVL;
      poolData.push({ token: "ANML", erthPrice: globalErthPrice, tvl: anmlTVL });
      console.log(`[DEBUG] ANML pool: anmlPrice=${anmlPriceFinal}, anmlTVL=${anmlTVL}, anmlMarketCap=${anmlMarketCap}`);
    }

    // 8) Create a single global data point including all pool data and total TVL, plus explicit ANML fields
    const now = new Date();
    // Set time to midnight (00:00:00) to store only the date part
    now.setHours(0, 0, 0, 0);

    const dataPoint = {
      timestamp: now.getTime(), // Use midnight timestamp to represent just the date
      erthPrice: globalErthPrice,
      erthTotalSupply,
      erthMarketCap: globalErthPrice * erthTotalSupply,
      tvl: totalLiquidity,
      pools: poolData,
      anmlPrice: anmlPriceFinal, // Explicit ANML price
      anmlTotalSupply, // Add ANML supply
      anmlMarketCap, // Add ANML market cap
    };

    // Add the new data point and trim history if needed
    analyticsHistory.push(dataPoint);
    saveAnalyticsData();
    console.log("[analyticsManager] Updated global analytics:", dataPoint);
  } catch (err) {
    console.error("[analyticsManager] Error updating analytics:", err);
  }
}

// Schedule next update to occur once per day at midnight
function scheduleNextUpdate() {
  const now = new Date();
  // Set target time to next midnight
  const tomorrow = new Date(now);
  tomorrow.setDate(now.getDate() + 1);
  tomorrow.setHours(0, 0, 0, 0);

  // Calculate delay until next midnight
  const delay = tomorrow.getTime() - now.getTime();

  console.log(
    `[analyticsManager] Scheduling next update in ${Math.floor(delay / (1000 * 60 * 60))} hours (at midnight)`
  );

  setTimeout(() => {
    const updateTime = new Date();
    console.log(`[analyticsManager] Running scheduled update at ${updateTime.toLocaleTimeString()}`);
    updateErthValues().then(() => {
      // Schedule the next update after this one completes
      scheduleNextUpdate();
    });
  }, delay);
}

// Initialize analytics: load stored data, check if update is needed, then schedule for next day
function initAnalytics(resetData = false) {
  if (resetData) {
    resetAnalyticsData();
  } else {
    loadAnalyticsData();
  }

  // Check if we need to update (if no data or last entry is older than 1 day)
  const shouldUpdate = () => {
    if (analyticsHistory.length === 0) return true;

    const lastEntry = analyticsHistory[analyticsHistory.length - 1];
    const lastUpdateTime = new Date(lastEntry.timestamp);
    const now = new Date();

    // Calculate difference in days
    const diffTime = Math.abs(now - lastUpdateTime);
    const diffDays = Math.floor(diffTime / (1000 * 60 * 60 * 24));

    console.log(`[analyticsManager] Last update was ${diffDays} day(s) ago`);
    return diffDays >= 1;
  };

  // Always update after a reset or if data is stale
  if (resetData || shouldUpdate()) {
    console.log("[analyticsManager] Updating analytics data");
    updateErthValues().then(() => {
      // After update, schedule next one for tomorrow
      scheduleNextUpdate();
    });
  } else {
    console.log("[analyticsManager] Analytics data is recent, scheduling next update for tomorrow");
    // Just schedule next update without updating now
    scheduleNextUpdate();
  }
}

// Endpoint helpers
function getLatestData() {
  return analyticsHistory.length ? analyticsHistory[analyticsHistory.length - 1] : null;
}

function getAllData() {
  return analyticsHistory;
}

// Export functions for use in your server
module.exports = {
  initAnalytics,
  getLatestData,
  getAllData,
  resetAnalyticsData,
};

File: server/analytics_manager.py
import os
import json
import time
import asyncio
import aiohttp
from typing import List, Dict, Any

# Import shared resources and configuration
import config
from dependencies import secret_client

# In-memory cache for analytics data
analytics_history: List[Dict[str, Any]] = []

# --- File Handling ---

def load_analytics_data():
    """
    Loads historical analytics data from the JSON file into the in-memory cache.
    If the file doesn't exist, it initializes an empty list.
    """
    global analytics_history
    try:
        if os.path.exists(config.ANALYTICS_FILE):
            with open(config.ANALYTICS_FILE, "r") as f:
                analytics_history = json.load(f)
                print(f"[Analytics] Loaded {len(analytics_history)} historical data points from {config.ANALYTICS_FILE}")
        else:
            analytics_history = []
            print(f"[Analytics] No analytics file found at {config.ANALYTICS_FILE}. Starting fresh.")
    except Exception as e:
        analytics_history = []
        print(f"[Analytics] ERROR: Could not load analytics data: {e}. Starting fresh.")

def save_analytics_data():
    """Saves the current in-memory analytics history to the JSON file."""
    try:
        with open(config.ANALYTICS_FILE, "w") as f:
            json.dump(analytics_history, f, indent=2)
            print(f"[Analytics] Successfully saved {len(analytics_history)} data points to {config.ANALYTICS_FILE}")
    except Exception as e:
        print(f"[Analytics] ERROR: Could not save analytics data: {e}")

# --- Core Analytics Logic ---

async def update_analytics_job():
    """
    The main job to be run by the scheduler. It fetches all necessary data
    from external APIs and on-chain contracts to build a new analytics data point.
    """
    print("[Analytics] Starting scheduled update...")
    try:
        # 1. Fetch prices for sSCRT and FINA from CoinGecko
        token_ids_to_fetch = [t["coingeckoId"] for t in config.TOKENS.values() if "coingeckoId" in t]
        coingecko_ids_str = ",".join(token_ids_to_fetch)
        price_url = f"https://api.coingecko.com/api/v3/simple/price?ids={coingecko_ids_str}&vs_currencies=usd"
        
        prices = {}
        async with aiohttp.ClientSession() as session:
            print(f"[Analytics] Fetching prices from CoinGecko for: {coingecko_ids_str}")
            async with session.get(price_url) as resp:
                if resp.status != 200:
                    raise Exception(f"CoinGecko API call failed with status {resp.status}: {await resp.text()}")
                price_data = await resp.json()
                print(f"[Analytics] CoinGecko response: {price_data}")

                # Map prices back to our token symbols (sSCRT, FINA)
                for symbol, token_info in config.TOKENS.items():
                    if "coingeckoId" in token_info and token_info["coingeckoId"] in price_data:
                        prices[symbol] = price_data[token_info["coingeckoId"]]["usd"]

        print(f"[Analytics] Fetched prices (USD): {prices}")

        # 2. Query token total supplies
        print("[Analytics] Querying ERTH and ANML total supply...")
        erth_info = secret_client.wasm.contract_query(config.TOKENS['ERTH']['contract'], {"token_info": {}})
        erth_total_supply = int(erth_info["token_info"]["total_supply"]) / (10**config.TOKENS['ERTH']['decimals'])

        anml_info = secret_client.wasm.contract_query(config.TOKENS['ANML']['contract'], {"token_info": {}})
        anml_total_supply = int(anml_info["token_info"]["total_supply"]) / (10**config.TOKENS['ANML']['decimals'])
        print(f"[Analytics] ERTH Supply: {erth_total_supply}, ANML Supply: {anml_total_supply}")

        # 3. Query the unified pool for reserves
        # Get contract addresses for all tokens except ERTH
        pool_addresses = [t["contract"] for k, t in config.TOKENS.items() if k != "ERTH"]
        print(f"[Analytics] Querying unified pool {config.UNIFIED_POOL_CONTRACT}...")
        unified_pool_res = secret_client.wasm.contract_query(
            config.UNIFIED_POOL_CONTRACT, {"query_pool_info": {"pools": pool_addresses}}
        )
        print(f"[Analytics] Unified pool response received.")

        # 4. Calculate ERTH price based on weighted TVL
        total_weighted_price = 0
        total_liquidity_for_erth_calc = 0
        all_pool_data = []
        anml_pool_reserves = None
        global_erth_price = 0

        # First pass: process pools with known prices (sSCRT, FINA) to determine ERTH price
        for i, pool_state in enumerate(unified_pool_res):
            # Determine which token this pool is for. Assumes a consistent order.
            # Skips ERTH (index 0) and maps to ANML, FINA, sSCRT
            token_symbol = list(config.TOKENS.keys())[i + 1]
            token_meta = config.TOKENS[token_symbol]
            
            erth_reserve = int(pool_state["state"]["erth_reserve"]) / (10**config.TOKENS['ERTH']['decimals'])
            token_reserve = int(pool_state["state"]["token_b_reserve"]) / (10**token_meta['decimals'])
            
            if token_symbol == "ANML":
                anml_pool_reserves = {"erth_reserve": erth_reserve, "token_reserve": token_reserve}
                continue # Skip ANML for now, we'll process it after we have ERTH price
            
            # For FINA and sSCRT pools
            token_price_usd = prices[token_symbol]
            pool_erth_price = (token_reserve / erth_reserve) * token_price_usd if erth_reserve else 0
            pool_tvl = (token_reserve * token_price_usd) + (erth_reserve * pool_erth_price)

            total_weighted_price += pool_erth_price * pool_tvl
            total_liquidity_for_erth_calc += pool_tvl

            all_pool_data.append({"token": token_symbol, "tvl": pool_tvl})

        if total_liquidity_for_erth_calc > 0:
            global_erth_price = total_weighted_price / total_liquidity_for_erth_calc
        print(f"[Analytics] Global ERTH Price calculated: ${global_erth_price:.6f}")

        # 5. Second pass: Calculate ANML price and TVL using the now-known ERTH price
        anml_price_usd = 0
        if anml_pool_reserves and anml_pool_reserves["token_reserve"] > 0:
            anml_price_usd = (anml_pool_reserves["erth_reserve"] / anml_pool_reserves["token_reserve"]) * global_erth_price
            anml_tvl = (anml_pool_reserves["token_reserve"] * anml_price_usd) + (anml_pool_reserves["erth_reserve"] * global_erth_price)
            all_pool_data.append({"token": "ANML", "tvl": anml_tvl})
        print(f"[Analytics] ANML Price calculated: ${anml_price_usd:.6f}")

        # 6. Assemble the final data point
        total_tvl = sum(p['tvl'] for p in all_pool_data)
        
        # Normalize timestamp to the beginning of the current day (UTC)
        now_utc_day_start = int(time.time() // 86400 * 86400 * 1000)

        data_point = {
            "timestamp": now_utc_day_start,
            "erthPrice": global_erth_price,
            "erthTotalSupply": erth_total_supply,
            "erthMarketCap": global_erth_price * erth_total_supply,
            "tvl": total_tvl,
            "pools": all_pool_data,
            "anmlPrice": anml_price_usd,
            "anmlTotalSupply": anml_total_supply,
            "anmlMarketCap": anml_price_usd * anml_total_supply,
        }

        # Avoid adding duplicate entries for the same day
        if any(p['timestamp'] == now_utc_day_start for p in analytics_history):
            print(f"[Analytics] Data for timestamp {now_utc_day_start} already exists. Skipping add.")
        else:
            analytics_history.append(data_point)
            print(f"[Analytics] New data point created: {data_point}")
            save_analytics_data()

    except Exception as e:
        print(f"[Analytics] FATAL ERROR during analytics update: {e}")
        # Optionally, re-raise the exception if you want the scheduler to log it as a job failure
        # raise e

# --- Initialization ---

def init_analytics():
    """
    Initializes the analytics manager. This should be called once on application startup.
    It loads historical data and runs an immediate update if the data is stale.
    """
    print("[Analytics] Initializing...")
    load_analytics_data()

    # Check if the last data point is older than 24 hours
    is_stale = True
    if analytics_history:
        last_timestamp_ms = analytics_history[-1]["timestamp"]
        # Convert ms to seconds for comparison
        if (time.time() - last_timestamp_ms / 1000) < 86400: # 24 * 60 * 60 seconds
            is_stale = False

    if is_stale:
        print("[Analytics] Data is stale or missing. Running initial update now.")
        # Run the async update function in a blocking way for startup
        asyncio.run(update_analytics_job())
    else:
        print("[Analytics] Data is up-to-date. Next update will be scheduled.")